src/cursor/common.rs

```diff
@@ -214,3 +246,63 @@ pub(crate) struct CursorInformation {
     pub(crate) batch_size: Option<u32>,
     pub(crate) max_time: Option<Duration>,
 }
+
+#[derive(Clone, Debug)]
+pub(crate) enum PinnedConnection {
+    Valid(PinHandle),
+    Invalid(PinHandle),
+    Unpinned,
+}
+
+impl PinnedConnection {
+    pub(super) fn new(handle: Option<PinHandle>) -> Self {
+        match handle {
+            Some(h) => Self::Valid(h),
+            None => Self::Unpinned,
+        }
+    }
+
+    fn handle(&self) -> Option<&PinHandle> {
```

Comment:
This method might be a little bit dangerous to use (see my comment above about stopping after invalidation) since we may only want to use the handle if the connection is thought to be valid. What about splitting this up into `valid_handle()` and `any_handle()` or something like that?

Reply:
I think we want to continue to provide the handle for a pinned connection even if it's invalidated; for `start_get_more` and `kill_cursor`, it would otherwise check a fresh connection out and potentially cause bad things to happen if the cursor id happened to collide with one on a different server behind the load balancer, and for `unpin` it doesn't matter if it's valid or not.

Reply:
Oh good point, my suggestion may have been even more dangerous...